### ISSUE TITLE
[docs] [python] custom metric function in `lgb.train()` interface should not refer to the the passed dataset a `train_data`

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#4759]
## Summary

This issue proposes changing the documentation for `feval` in `lgb.train()`, in the Python package, replacing `train_data` with `evaluation_dataset`.

## Motivation

`lgb.train()` allows providing custom evaluation metrics defined in Python functions. According to the docs for `lgb.train()` ([here](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.train.html#lightgbm.train)), such functions should have the following signature.

> Should accept two parameters: preds, train_data, and return (grad, hess).
>
> **train_data: Dataset**
>    The training dataset.

I think it's confusing to refer to this as "The training dataset". In my opinion, that gives the impression that custom evaluation functions will have access to the data being used for training, which is not true. Instead, each custom evaluation function is provided *a Dataset that evaluation has been request for*, which might be either the training data or a separate validation dataset.

https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/engine.py#L296-L299

https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/basic.py#L3240-L3255

https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/basic.py#L3271-L3272

## Description

Specifically, I'm proposing the following changes:

* change the following to `evaluation_data`, not `valid_data` https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/basic.py#L3254-L3255
* change the following to `evaluation_data`, not `train_data` https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/engine.py#L88-L89

## References

Noticed this while working on https://github.com/microsoft/LightGBM/issues/4679#issuecomment-955872028.

**Comment by jameslamb** [comment]
@StrikerRUS @jmoralez would you support a change like this?

**Comment by shiyu1994** [comment]
I agree with the changes.
In fact, the description of `feval` seems just copied from the description of `fobj`, which is used only by training dataset. And this line should also be changed:
https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/engine.py#L81

**Comment by StrikerRUS** [comment]
Generally, I support this change. But maybe use `eval_data` name instead of `evaluation_data` for the consistency with other similar names like ` eval_valid`, ` eval_train`, `eval_name`, `eval_result`, etc.?

**Comment by akshitadixit** [comment]
Is this available to work on?

**Comment by jameslamb** [comment]
@akshitadixit sure! We'd welcome the help!

Please see @StrikerRUS 's comment above about using the new name `eval_data`: https://github.com/microsoft/LightGBM/issues/4759#issuecomment-956205618

**Comment by akshitadixit** [comment]
Sure!

**Comment by akshitadixit** [comment]
> I agree with the changes. In fact, the description of `feval` seems just copied from the description of `fobj`, which is used only by training dataset. And this line should also be changed:
> 
> https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/engine.py#L81

there are lines in the description of `feval` similar this one that mention `train_data` instead of `eval_data`, do I change those too? for eg: 
https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/basic.py#L3214
https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/basic.py#L3221

**Comment by akshitadixit** [comment]
Hi @jameslamb and @StrikerRUS  i noticed something I though of confirming from you.

I have been asked to change `valid_data` to `eval_data` in the `eval_valid()` function
https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/basic.py#L3240-L3256

whereas there is a similar instance with `train_data` in the `eval_train()` function too as was pointed out for [engine.py](https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/engine.py):
https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/basic.py#L3207-L3223

Kindly clarify this,

**Comment by jameslamb** [comment]
@akshitadixit sorry for the delayed response.

Yes, please also change those places to

```text
Should accept two parameters: preds, eval_data
```
